google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/schema/SpannerSchemaManagementTool.java

```diff
@@ -34,6 +45,111 @@
  * DDL statements.
  */
 public class SpannerSchemaManagementTool extends HibernateSchemaManagementTool {
+  /**
+   * Custom implementation for {@link DdlTransactionIsolator} that will automatically create a proxy
+   * for Connection and Statement. These proxies will again be used to override the default behavior
+   * of the `START BATCH DDL` and `RUN BATCH` statements, by converting any {@link SQLException}
+   * that is returned by these methods into a {@link SpannerExceptionFactory}. The reason for this
+   * is that `START BATCH DDL` and `RUN BATCH` are added to each schema migration as auxiliary
+   * database objects automatically by the {@link SpannerSchemaManagementTool}. Hibernate will
+   * however silently ignore any {@link SQLException} that is thrown for auxiliary database objects.
+   * This means that if for example `RUN BATCH` fails, Hibernate will still report success for the
+   * entire migration. Throwing a {@link com.google.cloud.spanner.SpannerException} instead does
+   * cause an error to be returned for the migration.
+   */
+  static class SpannerDdlTransactionIsolator implements DdlTransactionIsolator {
+    private static final AbstractStatementParser PARSER =
+        AbstractStatementParser.getInstance(Dialect.GOOGLE_STANDARD_SQL);
+    private final Method createStatementMethod;
+    private final Method executeMethod;
+    private final DdlTransactionIsolator delegate;
+
+    SpannerDdlTransactionIsolator(DdlTransactionIsolator delegate) throws NoSuchMethodException {
+      this.delegate = delegate;
+      this.createStatementMethod = Connection.class.getDeclaredMethod("createStatement");
+      this.executeMethod = Statement.class.getDeclaredMethod("execute", String.class);
+    }
+
+    @Override
+    public JdbcContext getJdbcContext() {
+      return delegate.getJdbcContext();
+    }
+
+    @Override
+    public void prepare() {
+      delegate.prepare();
+    }
+
+    @Override
+    public Connection getIsolatedConnection() {
+      Connection delegateConnection = this.delegate.getIsolatedConnection();
+      // Create a proxy for the connection that will override the call to
+      // Connection#createStatement().
+      return (Connection)
+          Proxy.newProxyInstance(
+              delegateConnection.getClass().getClassLoader(),
+              new Class[] {Connection.class},
+              (proxy, method, args) -> {
+                // Only handle the Connection#createStatement() differently.
+                // All other methods are just passed through.
+                if (method.equals(createStatementMethod)) {
+                  // Create a proxy for the returned Statement that will override the behavior of
+                  // Statement#execute(String).
+                  Statement delegateStatement = delegateConnection.createStatement();
+                  return Proxy.newProxyInstance(
```

Comment:
Do you think it's better to extract the `return` statement to a separate method?

Reply:
Yes, it will make it more readable. Done.